app/views/questionnaire.py

```diff
@@ -151,22 +168,99 @@ def post_block(routing_path, schema, metadata, collection_metadata, answer_store
+        if next_url:
+            return redirect(next_url)
+
+    answer_store_updater.save_answers(form)
+    next_location = path_finder.get_next_location(current_location=current_location)
+
+    if _is_end_of_questionnaire(block, next_location):
+        return submit_answers(routing_path, schema)
+
+    return redirect(next_location.url())
+
+
+def perform_list_action(schema, metadata, answer_store, current_location, form, rendered_block, answer_store_updater, list_item_id):
+    list_collector_child = schema.is_block_list_collector_child(current_location.block_id)
+
+    block = schema.get_block(current_location.block_id)
+
+    parent_block = schema.get_list_collector_for_block_id(current_location.block_id)
+
+    if block['type'] == 'ListCollector':
         answer_store_updater.save_answers(form)
+        if list(form.data.values())[0] == block['add_answer_value']:
+            add_url = url_for('questionnaire.get_add_list_item', list_name=rendered_block['populates_list'], add_block_id=rendered_block['add_block']['id'])
+            return add_url
+    elif block['type'] == 'ListRemoveQuestion':
+        if list(form.data.values())[0] == parent_block['remove_answer_value']:
+            list_name = parent_block['populates_list']
+            answer_store_updater.remove_all_answers_with_list_item_id(list_name, list_item_id)
+        else:
+            return url_for('questionnaire.get_block', block_id=parent_block['id'])
+    elif block['type'] == 'ListAddQuestion':
+        answer_store_updater.add_new_list_item_answers(form, parent_block['populates_list'])
+    elif block['type'] == 'ListEditQuestion':
+        answer_store_updater.save_answers(form, save_completed_blocks=False)
+
+    if list_collector_child:
+        # Clear the answer from the confirmation question on the list collector question
+        transformed_parent = transform_variants(parent_block, schema, metadata, answer_store)
+        answer_ids_to_remove = [answer['id'] for answer in transformed_parent['question']['answers']]
+        answer_store_updater.remove_answer_ids(answer_ids_to_remove)
+
+        list_collector_url = url_for('questionnaire.get_block', block_id=parent_block['id'])
+
+        return list_collector_url
 
-        next_location = path_finder.get_next_location(current_location=current_location)
 
-        if _is_end_of_questionnaire(block, next_location):
-            return submit_answers(routing_path, schema)
+@questionnaire_blueprint.route('<block_id>', methods=['POST'])
+@login_required
+@with_answer_store
+@with_list_store
+@with_collection_metadata
+@with_metadata
+@with_schema
+@full_routing_path_required
+# pylint: disable=too-many-locals, too-many-return-statements
+def post_block(routing_path, schema, metadata, collection_metadata, list_store, answer_store, block_id):
+    return post_block_handler(routing_path, schema, metadata, collection_metadata, list_store, answer_store, block_id)
```

Comment:
I think we should consider a `@with_questionnaire store` decorator at the top level of routes, as currently the routes have individual `with_answer_store`, `with_list_store`, `with_collection_metadata`, `with_metadata` - all of which call `get_questionnaire_store`. `questionnaire_store` could then be passed to post_block_handler instead of `metadata, collection_metadata, list_store, answer_store`.

We can then only pass what's required to methods called from `post_block_handler` etc by accessing them from `questionnaire_store`, e.g. `questionnaire_store.answer_store`.